pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java

```diff
@@ -124,12 +132,26 @@ public void init(TableConfig tableConfig, Schema schema, Map<String, String> bat
         "batchConfigMaps must contain only 1 BatchConfig for table: %s", _tableNameWithType);
 
     Map<String, String> batchConfigMap = _batchIngestionConfig.getBatchConfigMaps().get(0);
+    batchConfigMap.put(BatchConfigProperties.PARTITION_ID, Integer.toString(_indexOfSubtask));
+    batchConfigMap.put(BatchConfigProperties.SEQUENCE_ID, Integer.toString(_seqId));
+    batchConfigMap.put(BatchConfigProperties.SEGMENT_CREATION_TIME_MS, String.valueOf(_segmentCreationTimeMs));
+    batchConfigMap.put(
+        BatchConfigProperties.SEGMENT_NAME_GENERATOR_PROP_PREFIX + "." + BatchConfigProperties.SEGMENT_NAME_PREFIX,
+        _segmentNamePrefix);
+
+    // generate segment name for simple segment name generator type(non upsert tables)
     String segmentNamePostfixProp = String.format("%s.%s", BatchConfigProperties.SEGMENT_NAME_GENERATOR_PROP_PREFIX,
         BatchConfigProperties.SEGMENT_NAME_POSTFIX);
     String segmentSuffix = batchConfigMap.get(segmentNamePostfixProp);
     segmentSuffix = segmentSuffix == null ? String.valueOf(_indexOfSubtask) : segmentSuffix + "_" + _indexOfSubtask;
     batchConfigMap.put(segmentNamePostfixProp, segmentSuffix);
 
+    // For upsert tables must use the UploadedRealtimeSegmentName for right assignment of segments
+    if (_tableConfig.isUpsertEnabled()) {
+      batchConfigMap.put(BatchConfigProperties.SEGMENT_NAME_GENERATOR_TYPE,
+          BatchConfigProperties.SegmentNameGeneratorType.UPLOADED_REALTIME);
+    }
+
```

Comment:
Not introduced in this PR: Since right assignment of segments is strictly required in upsert tables and not other non-upsert realtime tables, do you think SegmentNameGeneratorType.UPLOADED_UPSERT would have made more sense?

Reply:
The segment name though more useful for upsert is actually needed for all realtime tables to differentiate between stream generated segments and uploaded segments.

Reply:
Hmm in that case should we check if the table is of realtime nature rather than only upsert table?

Reply:
We ideally should have it. I wouldn't be sure if somebody has made this work for non upsert table with LLCSegmentName or another name format with a minor patch in server side and with default segment assignment strategy, in that case this change will break uploads.

So was the reason to mandate it for upserts only which were broken until https://github.com/apache/pinot/pull/13107

Reply:
In that case, they will have an internal implementation of `SegmentWriter` I guess. Anyways, if there are internal patches if people are using to make this working then better to have it in Pinot as first-class itself.

Reply:
Making `UploadedRealtimeSegmentName` as the only segment naming format supported to upload to a reltime table.